src/interface/state_descriptor.hpp

```diff
@@ -49,8 +49,11 @@ class StateDescriptor {
   explicit StateDescriptor(std::string const &label) : label_(label) {}
 
   template <typename T>
-  void AddParam(const std::string &key, T value) {
+  void AddParam(const std::string &key, T value, const bool include_with_output = false) {
```

Comment:
Is there a reason to not just include all parameters with the output and not give the user an option?

Reply:
That would be an interesting idea. We don't have that many parameters, so it wouldn't be that much of a file size change to write them all to disk. It might clutter the output a bit but it would simplify the code.

The one obstacle is that currently we don't have methods to write all types of parameters as an HDF5 attribute. The simple integer, floating point, and string types are easy to write, but more complicated ones like the equation of state class that we have in AthenaPK.

I think it would make sense to write all simple parameters in a separate `Params` dataset as long as we're ok with all that data in the HDF5 file. We could limit it to simple types for now and in the future let parameter types implement a `WriteAsHDF5Attribute` if they also need to be added to the HDF5 file. We don't have any use cases for that though.

For this PR I can make all `int`, `double`, and `std::string`  parameters write as an attribute and in your PR we can add other simple types.

Reply:
I realized this later. Yes, I think just writing all params with simple type to the outputfile and ignoring complex types is the right way to go.